nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/nar/StandardNarManager.java

```diff
@@ -0,0 +1,504 @@
+        return Optional.ofNullable(narNode);
+    }
+
+    @Override
+    public synchronized void verifyDeleteNar(final String identifier, final boolean forceDelete) {
+        final NarNode narNode = getNarNodeOrThrowNotFound(identifier);
+
+        // Always allow deletion of a NAR that is not fully installed
+        if (narNode.getState() != NarState.INSTALLED) {
+            return;
+        }
+
+        final BundleCoordinate coordinate = narNode.getManifest().getCoordinate();
+        final Set<Bundle> bundlesWithMatchingDependency = extensionManager.getDependentBundles(coordinate);
+        if (!bundlesWithMatchingDependency.isEmpty()) {
+            throw new IllegalStateException("Unable to delete NAR because it is a dependency of other NARs");
+        }
+
+        if (!forceDelete && narComponentManager.componentsExist(coordinate)) {
+            throw new IllegalStateException("Unable to delete NAR because components are instantiated from this NAR");
+        }
+    }
+
+    @Override
+    public synchronized NarNode deleteNar(final String identifier) throws IOException {
+        final NarNode narNode = getNarNodeOrThrowNotFound(identifier);
+        final BundleCoordinate coordinate = narNode.getManifest().getCoordinate();
+        LOGGER.info("Deleting NAR with id [{}] and coordinate [{}]", identifier, coordinate);
+
+        final Future<?> installTask = installFuturesById.remove(identifier);
+        if (installTask != null) {
+            installTask.cancel(true);
+        }
+
+        final Bundle existingBundle = extensionManager.getBundle(coordinate);
+        if (existingBundle != null) {
+            narLoader.unload(existingBundle);
+        }
+
+        deleteExecutorService.submit(() -> {
+            LOGGER.info("Unloading components for deleting NAR with id [{}] and coordinate [{}]", identifier, coordinate);
+            final StandardStoppedComponents stoppedComponents = new StandardStoppedComponents(controllerServiceProvider);
+            narComponentManager.unloadComponents(coordinate, stoppedComponents);
+            LOGGER.info("Completed unloading components for deleting NAR with id [{}] and coordinate [{}]", identifier, coordinate);
+        });
+
+        persistenceProvider.deleteNar(coordinate);
+        narNodesById.remove(identifier);
+
+        return narNode;
+    }
+
+    @Override
+    public synchronized InputStream readNar(final String identifier) {
+        final NarNode narNode = getNarNodeOrThrowNotFound(identifier);
+        final BundleCoordinate coordinate = narNode.getManifest().getCoordinate();
+        try {
+            return persistenceProvider.readNar(coordinate);
+        } catch (final FileNotFoundException e) {
+            throw new NarNotFoundException(coordinate);
+        }
+    }
+
+    @Override
+    public synchronized void syncWithClusterCoordinator() {
+        if (clusterCoordinator == null) {
+            LOGGER.info("Cluster coordinator is null, will not sync NARs");
```

Comment:
It seems like this should be changed to `debug`. Also recommend rewording:
```suggestion
            LOGGER.info("Cluster coordinator is configured: NAR synchronization disabled");
```